src/Sema.zig

```diff
@@ -35243,6 +35248,18 @@ fn resolveUnionLayout(sema: *Sema, ty: Type) CompileError!void {
         );
         return sema.failWithOwnedErrorMsg(null, msg);
     }
+
+    if (union_obj.flagsPtr(ip).assumed_pointer_aligned and
+        alignment.compareStrict(.neq, Alignment.fromByteUnits(@divExact(mod.getTarget().ptrBitWidth(), 8))))
+    {
+        const msg = try Module.ErrorMsg.create(
+            sema.gpa,
+            mod.declPtr(union_obj.decl).srcLoc(mod),
+            "union layout depends on being pointer aligned",
+            .{},
+        );
+        return sema.failWithOwnedErrorMsg(null, msg);
```

Comment:
It'd be great to have a compile error test for this if we can figure out how to reach it. Maybe the following works (completely untested):
```zig
const U = union {
    next: ?*align(1) U align(128),
};
```

Reply:
Good call, added `test/cases/union_depends_on_pointer_alignment.zig`.